### PR TITLE
Lwm2m client state machine update

### PR DESCRIPTION
**Pull request by jheiskan81**
LwM2M engine pause and resume was not safe to call at LTE event handler. 

**Changes:**

LwM2M client state machine updated that it handle LTE network registration status for suspending and Resume in same thread than RD client stop and Start.

Removed LTE notification from subsys and moved functionality to application for controlling LTE connection status.

### CODE REVIEW

**Comment by jheiskan81** [comment]
@SeppoTakalo @jarlamsa @VeijoPesonen @juhaylinen @rlubos Please review

File: samples/nrf9160/lwm2m_client/src/main.c [hunk 1]
```diff
@@ -57,19 +57,27 @@ LOG_MODULE_REGISTER(app_lwm2m_client, CONFIG_APP_LOG_LEVEL);
...
 
+/* Client State Machine states*/
+enum lwm2m_client_state {
+	START,		/* Start Connection to server*/
+	CONNECTING,	/* Client is connecting to server */
+	BOOTSTRAP,	/* Client is doing bootstrap */
+	CONNECTED,	/* Client connection establisment to server */
+	LTE_OFFLINE,	/* LTE offline and LwM2M engine suspended */
+	NETWORK_ERROR	/* Client networ error handling. Client stop and modem reset */
+};
```
Comment (SeppoTakalo) [inline samples/nrf9160/lwm2m_client/src/main.c:68]: There is confusion about what "client" is, so lets call the Zephyr side library as "engine" and application as a "client"

```suggestion
/* Client states*/
enum lwm2m_client_state {
	START,		/* Start Connection to a server*/
	CONNECTING,	/* LwM2M engine is connecting  */
	BOOTSTRAP,	/* LwM2M enginet is doing a bootstrap */
	CONNECTED,	/* LwM2M Client connection establisment to server */
	LTE_OFFLINE,	/* LTE offline and LwM2M engine should be suspended */
	NETWORK_ERROR	/* Client networ error handling. Client stop and modem reset */
};
```

Comment (jheiskan81) [reply samples/nrf9160/lwm2m_client/src/main.c:68]: Ah OK let me fix this.

**Comment by jheiskan81** [comment]
@SeppoTakalo Please review I fix your comment's

File: samples/nrf9160/lwm2m_client/src/main.c [hunk 2]
```diff
@@ -57,19 +57,27 @@ LOG_MODULE_REGISTER(app_lwm2m_client, CONFIG_APP_LOG_LEVEL);
...
+
+static enum lwm2m_client_state client_state = START;
+
 static uint8_t endpoint_name[ENDPOINT_NAME_LEN + 1];
 static uint8_t imei_buf[IMEI_LEN + sizeof("\r\nOK\r\n")];
 static struct lwm2m_ctx client = {0};
 static bool reconnect;
-static int reconnection_counter;
-static struct k_sem lwm2m_restart;
+static struct k_sem lwm2m_state_machine;
```
Comment (SeppoTakalo) [inline samples/nrf9160/lwm2m_client/src/main.c:76]: Also we could use `K_SEM_DEFINE(state_mutex, 0, 1);` so we don't need to call `k_sem_init()`

**Comment by SeppoTakalo** [comment]
I think the `lwm2m_` prefix is used too much in the application side. It is hard to see which ones are application specific static callbacks and which ones are public API.

File: samples/nrf9160/lwm2m_client/src/main.c [hunk 3]
```diff
@@ -428,27 +449,55 @@ static void modem_connect(void)
...
+static void suspend_lwm2m_engine(void)
+{
+	int ret;
+
+	lwm2m_network_state_trig_unlock(LTE_OFFLINE);
+	ret = lwm2m_engine_pause();
+	if (ret) {
+		LOG_ERR("LwM2M engine pause fail %d", ret);
+		reconnect = true;
+		lwm2m_network_state_trig_unlock(NETWORK_ERROR);
```
[the remark this answers is not in the document]
Comment (jheiskan81) [reply samples/nrf9160/lwm2m_client/src/main.c:498]: It is already done

**Comment by jheiskan81** [comment]
@SeppoTakalo Thanks for comments I fixed those.

File: samples/nrf9160/lwm2m_client/src/main.c [hunk 4]
```diff
@@ -57,19 +57,27 @@ LOG_MODULE_REGISTER(app_lwm2m_client, CONFIG_APP_LOG_LEVEL);
...
 #define LWM2M_SECURITY_NO_SEC 3
 
+/* Client State Machine states*/
+enum client_state {
+	START,		/* Start Connection to a server*/
+	CONNECTING,	/* LwM2M engine is connecting to server */
+	BOOTSTRAP,	/* LwM2M engine is doing a bootstrap */
+	CONNECTED,	/* LwM2M Client connection establisment to server */
+	LTE_OFFLINE,	/* LTE offline and LwM2M engine should be suspended */
+	NETWORK_ERROR	/* Client networ error handling. Client stop and modem reset */
```
Comment (simensrostad) [inline samples/nrf9160/lwm2m_client/src/main.c:67]:
```suggestion
	NETWORK_ERROR	/* Client network error handling. Client stop and modem reset */
```

File: samples/nrf9160/lwm2m_client/src/main.c [hunk 5]
```diff
@@ -57,19 +57,27 @@ LOG_MODULE_REGISTER(app_lwm2m_client, CONFIG_APP_LOG_LEVEL);
...
+};
```
Comment (simensrostad) [inline samples/nrf9160/lwm2m_client/src/main.c:68]:
```suggestion
} client_state = START;
```

File: samples/nrf9160/lwm2m_client/src/main.c [hunk 6]
```diff
@@ -57,19 +57,27 @@ LOG_MODULE_REGISTER(app_lwm2m_client, CONFIG_APP_LOG_LEVEL);
...
+
+static enum client_state client_state = START;
```
Comment (simensrostad) [inline samples/nrf9160/lwm2m_client/src/main.c:70]:
```suggestion
```

File: samples/nrf9160/lwm2m_client/src/main.c [hunk 7]
```diff
@@ -57,19 +57,27 @@ LOG_MODULE_REGISTER(app_lwm2m_client, CONFIG_APP_LOG_LEVEL);
 #define LWM2M_SECURITY_CERTIFICATE 2
 #define LWM2M_SECURITY_NO_SEC 3
 
+/* Client State Machine states*/
+enum client_state {
```
Comment (simensrostad) [inline samples/nrf9160/lwm2m_client/src/main.c:61]:
```suggestion
static enum client_state {
```

**Review by simensrostad — APPROVED** [review]
Some simplifications. I'm not comfortable enough to properly review these changes as I'm not very familiar with the sample 🚀 One optimization could be to depend on the PDN library to know when a valid IP has been obtained. Instead of depending on LTE registration messages from the modem. It should be more accurate.
https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/libraries/modem/pdn.html
https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/nrf9160/pdn/README.html

You will also get notifications on whether IPv6 has been given or not, which might not be the case in some networks. Especially when roaming. 

Some applications / samples has already started to make this transition.

**Comment by jheiskan81** [comment]
@simensrostad Thanks for good comment's I fixed those and rebased with main branch.

**Comment by jheiskan81** [comment]
> Some simplifications. I'm not comfortable enough to properly review these changes as I'm not very familiar with the sample rocket One optimization could be to depend on the PDN library to know when a valid IP has been obtained. Instead of depending on LTE registration messages from the modem. It should be more accurate. https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/libraries/modem/pdn.html https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/nrf9160/pdn/README.html
> 
> You will also get notifications on whether IPv6 has been given or not, which might not be the case in some networks. Especially when roaming.
> 
> Some applications / samples has already started to make this transition.

Intresting point. @SeppoTakalo What do you think? I think this will give more details. Also if IP address is updated not just loosed.

**Comment by jheiskan81** [comment]
I have been tested PR and it fix the problem when face network rush which cause that we start discovering network cell and re-join. That discover rejoin process happen multiple time in 1 minute time period. Now LwM2M pause and resume works on synch and Client recover nicely that.
Is there still any comment's for this PR?

**Review by rlubos — APPROVED** [review]

**Pull request merged**